packages/block-editor/src/components/block-navigation/tree.js

```diff
@@ -73,20 +80,33 @@ export default function BlockNavigationTree( {
 			__experimentalPersistentListViewFeatures,
 			blockDropTarget,
 			isTreeGridMounted: isMounted.current,
+			expandedState,
+			setExpandedState,
 		} ),
 		[
 			__experimentalFeatures,
 			__experimentalPersistentListViewFeatures,
 			blockDropTarget,
 			isMounted.current,
+			expandedState,
+			setExpandedState,
 		]
 	);
 
+	const toggleExpandCollapse = ( clientId, activeRow ) => {
+		// TODO: expandedState is stale in the callback, we could alternatively use a data store here instead of context.
```

Comment:
The main thing that sticks out is state management and keyboard/focus handling. Right now keyboard handling is implemented at the Tree Grid level, which makes triggering callbacks at the branch level very difficult. I couldn't think of a good way of preserving nested collapsed state without resorting to a heavier data solution like a data store, or something a bit hacky like the click() call I made. Let me know if folks had thoughts around that.

Reply:
Oh hmm, 🤔 maybe I can see if using React `useReducer` helps here at all.